File: algorithms/balanced_binary_tree.c
// https://leetcode.com/problems/balanced-binary-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool helper(struct TreeNode *root, int *height) {
  if (!root) {
    *height = 0;
    return true;
  }

  int lh, rh;
  bool left = helper(root->left, &lh);
  bool right = helper(root->right, &rh);
  if (!left || !right) {
    return false;
  }

  if (abs(lh - rh) > 1) {
    return false;
  }

  *height = lh < rh ? rh + 1 : lh + 1;
  return true;
}

bool isBalanced(struct TreeNode *root) {
  if (!root) {
    return true;
  }

  int h = 0;
  return helper(root, &h);
}

File: algorithms/easy/balanced_binary_tree.kt
/**
 * Example: var ti = TreeNode(5) var v = ti.`val` Definition for a binary tree node. class
 * TreeNode(var `val`: Int) { var left: TreeNode? = null var right: TreeNode? = null }
 */
class Solution {
  fun depth(root: TreeNode?): Int {
    if (root == null) {
      return 0
    }

    val left = depth(root?.left)
    val right = depth(root?.right)
    if (left == -1 || right == -1) {
      return -1
    }

    if (abs(left - right) < 2) {
      return max(left, right) + 1
    }

    return -1
  }

  fun isBalanced(root: TreeNode?): Boolean {
    return depth(root) != -1
  }
}
